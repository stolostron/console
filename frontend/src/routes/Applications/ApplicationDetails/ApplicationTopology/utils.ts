/* Copyright Contributors to the Open Cluster Management project */

// Utility function to safely get nested properties (replaces lodash get)
export function safeGet<T = any>(obj: any, path: string | string[], defaultValue?: T): T {
  if (!obj || typeof obj !== 'object') return defaultValue as T

  const keys = Array.isArray(path) ? path : path.split(/[.[\]]+/).filter((key) => key !== '')
  let result = obj

  for (const key of keys) {
    if (result == null || typeof result !== 'object') {
      return defaultValue as T
    }
    result = result[key]
  }

  return result === undefined ? (defaultValue as T) : result
}

/**
 * Helper function to safely get nested properties from an object
 */
export const getNestedProperty = (obj: any, path: string | number | (string | number)[], defaultValue?: any): any => {
  if (obj == null) return defaultValue

  const pathArray = Array.isArray(path) ? path : String(path).split('.')
  let result = obj

  for (const key of pathArray) {
    if (result == null) return defaultValue
    result = result[key]
  }

  return result !== undefined ? result : defaultValue
}

// Utility function to safely set nested properties (replaces lodash set)
export function safeSet(obj: any, path: string | string[], value: any): void {
  if (!obj || typeof obj !== 'object') return

  const keys = Array.isArray(path) ? path : path.split('.')
  let current = obj

  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i]
    if (!(key in current) || typeof current[key] !== 'object' || current[key] === null) {
      current[key] = {}
    }
    current = current[key]
  }

  current[keys[keys.length - 1]] = value
}

// Deep clone utility (replaces lodash cloneDeep)
export function deepClone<T>(obj: T): T {
  if (obj === null || typeof obj !== 'object') return obj
  if (obj instanceof Date) return new Date(obj.getTime()) as unknown as T
  if (obj instanceof Array) return obj.map((item) => deepClone(item)) as unknown as T
  if (typeof obj === 'object') {
    const cloned = {} as T
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        cloned[key] = deepClone(obj[key])
      }
    }
    return cloned
  }
  return obj
}

/**
 * Deep equality comparison for objects
 * @param obj1 - First object to compare
 * @param obj2 - Second object to compare
 * @returns true if objects are deeply equal, false otherwise
 */
export function deepEqual(obj1: any, obj2: any): boolean {
  if (obj1 === obj2) return true
  if (obj1 == null || obj2 == null) return false
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false

  const keys1 = Object.keys(obj1)
  const keys2 = Object.keys(obj2)

  if (keys1.length !== keys2.length) return false

  for (const key of keys1) {
    if (!keys2.includes(key)) return false
    if (!deepEqual(obj1[key], obj2[key])) return false
  }

  return true
}
