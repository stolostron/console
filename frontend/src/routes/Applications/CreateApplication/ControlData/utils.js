export const getUniqueChannelName = (channelPath, groupControlData) => {
    //create a unique name for a new channel, based on path and type
    if (!channelPath || !groupControlData) {
      return ''
    }
  
    //get the channel type and append to url to make sure different type of channels are unique, yet using the same url
    const channelTypeSection = groupControlData.find(
      ({ id }) => id === 'channelType'
    )
  
    let channelTypeStr
    let channelType
    if (channelTypeSection) {
      channelTypeStr = _.get(channelTypeSection, 'active', [''])[0]
    }
  
    switch (channelTypeStr) {
    case 'github':
      channelType = 'g'
      break
    case 'helmrepo':
      channelType = 'h'
      break
    case 'objectstore':
      channelType = 'o'
      break
    default:
      channelType = 'ns'
    }
  
    let channelName = _.trim(channelPath)
    if (_.startsWith(channelName, 'https://')) {
      channelName = _.trimStart(channelName, 'https://')
    }
    if (_.startsWith(channelName, 'http://')) {
      channelName = _.trimStart(channelName, 'http://')
    }
    if (_.endsWith(channelName, '.git')) {
      channelName = _.trimEnd(channelName, '.git')
    }
  
    channelName = _.replace(channelName, /\./g, '')
    channelName = _.replace(channelName, /:/g, '')
    channelName = _.replace(channelName, /\//g, '-')
  
    channelName = _.trimEnd(channelName, '-')
    channelName = channelName.toLowerCase()
  
    //max name for ns or resources is 63 chars
    // trim channel name to max 58 char to allow a max of 63 char length
    //for the channel authentication (which is channelName-auth) object and channel ns (channelName-ns)
    if (channelName.length > 58) {
      channelName = channelName.substring(channelName.length - 56)
    }
    channelName = `${channelType}${channelName}`
  
    return channelName
  }