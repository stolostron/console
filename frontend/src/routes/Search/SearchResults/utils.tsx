/* Copyright Contributors to the Open Cluster Management project */
import { get } from 'lodash'
import queryString from 'query-string'
import { useContext, useMemo } from 'react'
import { TFunction } from 'react-i18next'
import { generatePath, NavigateFunction, useNavigate } from 'react-router-dom-v5-compat'
import { useTranslation } from '../../../lib/acm-i18next'
import { NavigationPath } from '../../../NavigationPath'
import { Cluster, fetchRetry, getBackendUrl } from '../../../resources'
import { useRecoilValue, useSharedAtoms } from '../../../shared-recoil'
import { AcmToastContext, compareStrings, IAlertContext } from '../../../ui-components'
import { useAllClusters } from '../../Infrastructure/Clusters/ManagedClusters/components/useAllClusters'
import {
  ClosedDeleteExternalResourceModalProps,
  IDeleteExternalResourceModalProps,
} from '../components/Modals/DeleteExternalResourceModal'
import { ClosedDeleteModalProps, IDeleteModalProps } from '../components/Modals/DeleteResourceModal'
import { searchClient } from '../search-sdk/search-client'
import { SearchResultItemsQuery } from '../search-sdk/search-sdk'
import { GetUrlSearchParam, SearchColumnDefinition } from '../searchDefinitions'

export interface ISearchResult {
  kind: string
  apiversion: string
  name: string
  apigroup?: string
  __type: string
}

export function handleVMActions(
  action: string,
  path: string,
  item: any,
  refetchVM: () => void, // provide a callback fn to refetch the vm
  toast: IAlertContext,
  t: TFunction
) {
  const abortController = new AbortController()
  fetchRetry({
    method: 'PUT',
    url: `${getBackendUrl()}${path}`,
    data: {
      managedCluster: item.cluster,
      vmName: item.name,
      vmNamespace: item.namespace,
    },
    signal: abortController.signal,
    retries: process.env.NODE_ENV === 'production' ? 2 : 0,
    headers: { Accept: '*/*' },
    disableRedirectUnauthorizedLogin: true,
  })
    .then(() => {
      // Wait 5 seconds to allow search collector to catch up & refetch search results to update table.
      setTimeout(refetchVM, 5000)
    })
    .catch((err) => {
      console.error(`VirtualMachine: ${item.name} ${action} error. ${err}`)

      let errMessage: string = err?.message ?? t('An unexpected error occurred.')
      if (errMessage.includes(':')) errMessage = errMessage.split(':').slice(1).join(':')
      if (errMessage === 'Unauthorized') errMessage = t('Unauthorized to execute this action.')
      toast.addAlert({
        title: t('Error triggering action {{action}} on VirtualMachine {{name}}', {
          name: item.name,
          action,
        }),
        message: errMessage,
        type: 'danger',
      })
    })
}

export const useGetRowActions = (
  resourceKind: string,
  currentQuery: string,
  relatedResource: boolean,
  setDeleteResource: React.Dispatch<React.SetStateAction<IDeleteModalProps>>,
  setDeleteExternalResource: React.Dispatch<React.SetStateAction<IDeleteExternalResourceModalProps>>
) => {
  const { t } = useTranslation()
  const navigate = useNavigate()
  const { settingsState } = useSharedAtoms()
  const vmActionsEnabled = useRecoilValue(settingsState)?.VIRTUAL_MACHINE_ACTIONS === 'enabled'
  const toast = useContext(AcmToastContext)
  const allClusters = useAllClusters(true)

  return useMemo(
    () =>
      getRowActions(
        resourceKind,
        currentQuery,
        relatedResource,
        setDeleteResource,
        setDeleteExternalResource,
        allClusters,
        navigate,
        toast,
        vmActionsEnabled,
        t
      ),
    [
      resourceKind,
      currentQuery,
      relatedResource,
      setDeleteResource,
      setDeleteExternalResource,
      allClusters,
      navigate,
      toast,
      vmActionsEnabled,
      t,
    ]
  )
}

export function getRowActions(
  resourceKind: string,
  currentQuery: string,
  relatedResource: boolean,
  setDeleteResource: React.Dispatch<React.SetStateAction<IDeleteModalProps>>,
  setDeleteExternalResource: React.Dispatch<React.SetStateAction<IDeleteExternalResourceModalProps>>,
  allClusters: Cluster[],
  navigate: NavigateFunction,
  toast: IAlertContext,
  vmActionsEnabled: boolean,
  t: TFunction
) {
  const viewApplication = {
    id: 'view-application',
    title: t('View Application'),
    click: (item: any) => {
      const { apigroup, applicationSet, cluster, name, namespace, kind, _hubClusterResource } = item
      if (apigroup === 'app.k8s.io' || apigroup === 'argoproj.io') {
        const path = generatePath(NavigationPath.applicationOverview, {
          namespace,
          name,
        })
        const params = queryString.stringify({
          apiVersion: `${kind}.${apigroup}`.toLowerCase(),
          cluster: !_hubClusterResource ? cluster : undefined,
          applicationset: applicationSet ?? undefined,
        })
        if (item.managedHub === 'global-hub' && !item?._hubClusterResource) {
          const hubUrl = allClusters.find((cluster) => cluster.name === item.cluster)?.consoleURL
          return window.open(`${hubUrl}${path}?${params}`, '_blank')
        }
        return navigate(
          {
            pathname: path,
            search: `?${params}`,
          },
          {
            state: {
              from: NavigationPath.search,
              fromSearch: window.location.search,
            },
          }
        )
      }
      const searchParams = GetUrlSearchParam(item)
      return navigate(
        {
          pathname: NavigationPath.resources,
          search: searchParams,
        },
        {
          state: {
            from: NavigationPath.search,
            fromSearch: window.location.search,
          },
        }
      )
    },
  }
  const viewAppTopology = {
    id: 'view-application-topology',
    title: t('View Application topology'),
    click: (item: any) => {
      const apiversion = encodeURIComponent(`${item?.kind}.${item?.apigroup}`.toLowerCase())
      const path = generatePath(NavigationPath.applicationTopology, { name: item.name, namespace: item.namespace })
      if (item.managedHub && !item?._hubClusterResource) {
        const hubUrl = allClusters.find((cluster) => cluster.name === item.cluster)?.consoleURL
        return window.open(`${hubUrl}${path}?apiVersion=${apiversion}`, '_blank')
      }
      return navigate(
        {
          pathname: path,
          search: `?apiVersion=${apiversion}`,
        },
        {
          state: {
            from: NavigationPath.search,
            fromSearch: window.location.search,
          },
        }
      )
    },
  }
  const editButton = {
    id: 'edit',
    title: t('Edit {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const searchParams = GetUrlSearchParam(item)
      if (item.managedHub && item.managedHub !== 'global-hub') {
        // If resource lives on a cluster managed by a managed hub we need to launch user to the managed hub for actions / viewing
        const hubUrl = allClusters.find((cluster) => cluster.name === item.managedHub)?.consoleURL
        return window.open(`${hubUrl}${NavigationPath.resourceYAML}${searchParams}`, '_blank')
      }
      return navigate(
        {
          pathname: NavigationPath.resourceYAML,
          search: searchParams,
        },
        {
          state: {
            from: NavigationPath.search,
            fromSearch: window.location.search,
          },
        }
      )
    },
  }
  const viewRelatedButton = {
    id: 'view-related',
    title: t('View related resources'),
    click: (item: any) => {
      const searchParams = GetUrlSearchParam(item)
      if (item.managedHub && item.managedHub !== 'global-hub') {
        const hubUrl = allClusters.find((cluster) => cluster.name === item.managedHub)?.consoleURL
        return window.open(`${hubUrl}${NavigationPath.resourceRelated}${GetUrlSearchParam(item)}`, '_blank')
      }
      return navigate(
        {
          pathname: NavigationPath.resourceRelated,
          search: searchParams,
        },
        {
          state: {
            from: NavigationPath.search,
            fromSearch: window.location.search,
          },
        }
      )
    },
  }
  const deleteButton = {
    id: 'delete',
    title: t('Delete {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      item.managedHub && item.managedHub !== 'global-hub'
        ? setDeleteExternalResource({
            open: true,
            close: () => setDeleteExternalResource(ClosedDeleteExternalResourceModalProps),
            resource: item,
            hubCluster: allClusters.find((cluster) => cluster.name === item.managedHub),
          })
        : setDeleteResource({
            open: true,
            close: () => setDeleteResource(ClosedDeleteModalProps),
            resource: item,
            isHubClusterResource: item._hubClusterResource === 'true',
            currentQuery,
            relatedResource,
          })
    },
  }

  const startVM = {
    id: 'startVM',
    title: t('Start {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const path = item?._hubClusterResource
        ? `/apis/subresources.kubevirt.io/v1/namespaces/${item.namespace}/virtualmachines/${item.name}/start`
        : `/virtualmachines/start`
      handleVMActions(
        'start',
        path,
        item,
        () => searchClient.refetchQueries({ include: ['searchResultItems'] }),
        toast,
        t
      )
    },
  }
  const stopVM = {
    id: 'stopVM',
    title: t('Stop {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const path = item?._hubClusterResource
        ? `/apis/subresources.kubevirt.io/v1/namespaces/${item.namespace}/virtualmachines/${item.name}/stop`
        : `/virtualmachines/stop`
      handleVMActions(
        'stop',
        path,
        item,
        () => searchClient.refetchQueries({ include: ['searchResultItems'] }),
        toast,
        t
      )
    },
  }
  const restartVM = {
    id: 'restartVM',
    title: t('Restart {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const path = item?._hubClusterResource
        ? `/apis/subresources.kubevirt.io/v1/namespaces/${item.namespace}/virtualmachines/${item.name}/restart`
        : `/virtualmachines/restart`
      handleVMActions(
        'restart',
        path,
        item,
        () => searchClient.refetchQueries({ include: ['searchResultItems'] }),
        toast,
        t
      )
    },
  }
  const pauseVM = {
    id: 'pauseVM',
    title: t('Pause {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const path = item?._hubClusterResource
        ? `/apis/subresources.kubevirt.io/v1/namespaces/${item.namespace}/virtualmachineinstances/${item.name}/pause`
        : `/virtualmachineinstances/pause`
      handleVMActions(
        'pause',
        path,
        item,
        () => searchClient.refetchQueries({ include: ['searchResultItems'] }),
        toast,
        t
      )
    },
  }
  const unpauseVM = {
    id: 'unpauseVM',
    title: t('Unpause {{resourceKind}}', { resourceKind }),
    click: (item: any) => {
      const path = item?._hubClusterResource
        ? `/apis/subresources.kubevirt.io/v1/namespaces/${item.namespace}/virtualmachineinstances/${item.name}/unpause`
        : `/virtualmachineinstances/unpause`
      handleVMActions(
        'unpause',
        path,
        item,
        () => searchClient.refetchQueries({ include: ['searchResultItems'] }),
        toast,
        t
      )
    },
  }

  if (
    resourceKind.toLowerCase() === 'cluster' ||
    resourceKind.toLowerCase() === 'release' ||
    resourceKind.toLowerCase() === 'policyreport'
  ) {
    return []
  } else if (resourceKind.toLowerCase() === 'application') {
    return [viewApplication, viewAppTopology, editButton, viewRelatedButton, deleteButton]
  } else if (resourceKind.toLowerCase() === 'virtualmachine') {
    return vmActionsEnabled
      ? [
          startVM,
          stopVM,
          restartVM,
          pauseVM,
          unpauseVM,
          { ...editButton, addSeparator: true },
          viewRelatedButton,
          deleteButton,
        ]
      : [editButton, viewRelatedButton, deleteButton]
  }
  return [editButton, viewRelatedButton, deleteButton]
}

export function generateSearchResultExport(
  exportFileName: string,
  searchResultData: SearchResultItemsQuery | undefined,
  searchDefinitions: any,
  toastContext: IAlertContext,
  t: TFunction<string, undefined>
) {
  toastContext.addAlert({
    title: t('Generating data. Download may take a moment to start.'),
    type: 'info',
    autoClose: true,
  })

  const searchResultItems: ISearchResult[] = searchResultData?.searchResult?.[0]?.items || []
  const kindSearchResultItems: Record<string, ISearchResult[]> = {}
  for (const searchResultItem of searchResultItems) {
    const apiGroup = searchResultItem?.apigroup
    const groupAndKind = apiGroup ? `${searchResultItem.kind}.${apiGroup}` : searchResultItem.kind
    const existing = kindSearchResultItems[groupAndKind]
    if (!existing) {
      kindSearchResultItems[groupAndKind] = [searchResultItem]
    } else {
      kindSearchResultItems[groupAndKind].push(searchResultItem)
    }
  }
  // Sort kinds alphabetically by kind then apigroup if same kind
  const kinds = Object.keys(kindSearchResultItems).sort((a, b) => {
    const strCompareRes = compareStrings(kindSearchResultItems[a][0].kind, kindSearchResultItems[b][0].kind)
    const getApiGroup = (type: string) =>
      kindSearchResultItems[type][0]?.apigroup
        ? `${kindSearchResultItems[type][0]?.apigroup}/${kindSearchResultItems[type][0]?.apiversion}`
        : ''
    return strCompareRes !== 0 ? strCompareRes : compareStrings(getApiGroup(a), getApiGroup(b))
  })

  // Sort resources alphabetically by name
  kinds.forEach((kind: string) => {
    kindSearchResultItems[kind] = kindSearchResultItems[kind].sort((a, b) => {
      return compareStrings(a.name, b.name)
    })
  })

  const defaulColumns = ['name', 'namespace', 'kind', 'cluster', 'created', 'label']
  let columns: SearchColumnDefinition[] = defaulColumns.map((defaultCol: string) => {
    return {
      header: defaultCol,
      sort: defaultCol,
      cell: defaultCol,
    }
  })

  // If only 1 resource kind -> use column headers from searchDefinitions
  if (kinds.length === 1) {
    const kindAndGroup =
      kinds[0].split('.')[0].toLowerCase() === 'subscription'
        ? `subscription.${searchResultItems[0].apigroup}`
        : kinds[0].split('.')[0].toLowerCase()
    columns = get(searchDefinitions, `['${kindAndGroup}'].columns`, searchDefinitions['genericresource'].columns)
    // Filter column definitions that do NOT contain sort field. Sort is the only way to confidently get resource fields
    columns = columns.filter((col: SearchColumnDefinition) => {
      return !!col.sort
    })
  }

  // Variable to store the final csv data
  const csv_data: string[] = [`${columns.map((col) => col.sort).join(',')}`]
  kinds.forEach((kind: string) => {
    kindSearchResultItems[kind].forEach((item: any) => {
      const csv_row: string[] = []
      columns.forEach((column: SearchColumnDefinition) => {
        csv_row.push(item[column.sort!] ?? '-') // Columns without a sort are filtered above
      })
      // Combine each column value with comma
      csv_data.push(csv_row.join(','))
    })
  })
  // Combine each row data with new line character
  const csv_string = csv_data.join('\n')

  // Create download
  const CSVFile = new Blob([csv_string], { type: 'text/csv' })
  const temp_link = document.createElement('a')
  temp_link.download = exportFileName
  const url = window.URL.createObjectURL(CSVFile)
  temp_link.href = url
  // This link should not be displayed
  temp_link.style.display = 'none'
  document.body.appendChild(temp_link)
  // Automatically click the link to trigger download
  temp_link.click()
  document.body.removeChild(temp_link)

  toastContext.addAlert({
    title: t('Export successful'),
    type: 'success',
    autoClose: true,
  })
}
