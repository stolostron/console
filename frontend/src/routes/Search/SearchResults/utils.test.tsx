// Copyright (c) 2023 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import i18next from 'i18next'
import { ClusterStatus } from '../../../resources'
import { getSearchDefinitions } from '../searchDefinitions'
import { generateSearchResultExport, getRowActions } from './utils'

const mockHistoryPush = jest.fn()
jest.mock('react-router-dom-v5-compat', () => ({
  ...jest.requireActual('react-router-dom-v5-compat'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))
const t = i18next.t.bind(i18next)
const navigate = jest.fn()
const toastContextMock: any = {
  addAlert: jest.fn(),
}

jest.mock('../../../resources', () => ({
  ...jest.requireActual('../../../resources'),
  fetchRetry: jest.fn(() => {
    return Promise.resolve()
  }),
}))
jest.mock('./utils', () => ({
  ...jest.requireActual('./utils'),
  handleVMActions: jest.fn(() => {
    return Promise.resolve()
  }),
}))

const allClusters = [
  {
    name: 'local-cluster',
    namespace: 'local-cluster',
    consoleURL: 'https://local-cluster.com',
    uid: '',
    status: ClusterStatus.ready,
    hasAutomationTemplate: false,
    hive: {
      isHibernatable: false,
    },
    isHive: false,
    isManaged: true,
    isCurator: false,
    isHostedCluster: false,
    isRegionalHubCluster: true,
    owner: {},
    isSNOCluster: false,
    isHypershift: false,
  },
  {
    name: 'leaf-hub',
    namespace: 'leaf-hub',
    consoleURL: 'https://leaf-hub.com',
    uid: '',
    status: ClusterStatus.ready,
    hasAutomationTemplate: false,
    hive: {
      isHibernatable: false,
    },
    isHive: false,
    isManaged: true,
    isCurator: false,
    isHostedCluster: false,
    isRegionalHubCluster: true,
    owner: {},
    isSNOCluster: false,
    isHypershift: false,
  },
]

test('Correctly return row Actions', () => {
  const vmActionsEnabled = false
  const res = getRowActions(
    'Pod',
    'kind:Pod',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  res[0].click({ kind: 'Pod' }) // edit resource
  res[1].click({ kind: 'Pod' }) // view related resources
  res[2].click({ kind: 'Pod' }) // delete resource
  expect(res).toMatchSnapshot()
})

test('Correctly return empty row Actions for restricted resource', () => {
  const vmActionsEnabled = false
  const res = getRowActions(
    'Cluster',
    'kind:Cluster',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  expect(res).toMatchSnapshot()
})

test('Correctly return empty row Actions for Application', () => {
  const vmActionsEnabled = false
  const res = getRowActions(
    'Application',
    'kind:Application',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  res[0].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'local-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
  }) // edit app
  res[1].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'local-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
  }) // view app topology
  expect(res).toMatchSnapshot()
})

test('Correctly return row Actions for Application in global search', () => {
  const vmActionsEnabled = false
  const res = getRowActions(
    'Application',
    'kind:Application',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  res[0].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'leaf-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
    managedHub: 'global-hub',
  }) // edit app
  res[1].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'leaf-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
    managedHub: 'global-hub',
  }) // view app topology
  res[2].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'leaf-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
    managedHub: 'global-hub',
  }) // edit app
  res[3].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'leaf-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
    managedHub: 'global-hub',
  }) // view related app resources
  res[4].click({
    apigroup: 'app.k8s.io',
    kind: 'Application',
    cluster: 'leaf-cluster',
    name: 'testApp',
    namespace: 'testAppNs',
    managedHub: 'global-hub',
  }) // delete app
  expect(res).toMatchSnapshot()
})

test('Correctly return VirtualMachine with actions enabled', () => {
  const vmActionsEnabled = true
  const res = getRowActions(
    'VirtualMachine',
    'kind:VirtualMachine',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  expect(res).toMatchSnapshot()
})

test('Correctly return VirtualMachine with actions disabled', () => {
  const vmActionsEnabled = false
  const res = getRowActions(
    'VirtualMachine',
    'kind:VirtualMachine',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  expect(res).toMatchSnapshot()
})
test('should handle vm action buttons', () => {
  const item = { managedHub: 'cluster1' }
  const vmActionsEnabled = true
  const actions = getRowActions(
    'VirtualMachine',
    'kind:VirtualMachine',
    false,
    () => {},
    () => {},
    allClusters,
    navigate,
    toastContextMock,
    vmActionsEnabled,
    t
  )
  const startVMAction = actions.find((action) => action.id === 'startVM')
  const stopVMAction = actions.find((action) => action.id === 'stopVM')
  const restartVMAction = actions.find((action) => action.id === 'restartVM')
  const pauseVMAction = actions.find((action) => action.id === 'pauseVM')
  const unpauseVMAction = actions.find((action) => action.id === 'unpauseVM')

  startVMAction?.click(item)
  stopVMAction?.click(item)
  restartVMAction?.click(item)
  pauseVMAction?.click(item)
  unpauseVMAction?.click(item)
})

test('generateSearchResultExport - Correctly generates and triggers csv download for single resource kind', () => {
  const toastContextMock: any = {
    addAlert: jest.fn(),
  }
  const t = (key: string) => key
  window.URL.createObjectURL = jest.fn()

  const searchResultDataMock = {
    searchResult: [
      {
        items: [
          {
            apigroup: 'operators.coreos.com',
            apiversion: 'v1alpha1',
            channel: 'release-2.11',
            cluster: 'local-cluster',
            created: '2024-04-15T14:20:26Z',
            installplan: 'advanced-cluster-management.v2.11.0',
            kind: 'Subscription',
            kind_plural: 'subscriptions',
            label: 'operators.coreos.com/advanced-cluster-management.open-cluster-management=',
            name: 'acm-operator-subscription',
            namespace: 'open-cluster-management',
            package: 'advanced-cluster-management',
            phase: 'AtLatestKnown',
            source: 'acm-custom-registry',
          },
        ],
        __typename: 'SearchResult',
      },
    ],
  }

  const searchDefinitions = getSearchDefinitions((key) => key)
  generateSearchResultExport('test-search-export', searchResultDataMock, searchDefinitions, toastContextMock, t)

  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Generating data. Download may take a moment to start.',
    type: 'info',
    autoClose: true,
  })
  expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1)
  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Export successful',
    type: 'success',
    autoClose: true,
  })
})

test('generateSearchResultExport - Correctly generates and triggers csv download for multi resource kind', () => {
  const toastContextMock: any = {
    addAlert: jest.fn(),
  }
  const t = (key: string) => key
  window.URL.createObjectURL = jest.fn()

  const searchResultDataMock = {
    searchResult: [
      {
        items: [
          {
            apigroup: 'operators.coreos.com',
            apiversion: 'v1alpha1',
            channel: 'release-2.11',
            cluster: 'local-cluster',
            created: '2024-04-15T14:20:26Z',
            installplan: 'advanced-cluster-management.v2.11.0',
            kind: 'Subscription',
            kind_plural: 'subscriptions',
            label: 'operators.coreos.com/advanced-cluster-management.open-cluster-management=',
            name: 'acm-operator-subscription',
            namespace: 'open-cluster-management',
            package: 'advanced-cluster-management',
            phase: 'AtLatestKnown',
            source: 'acm-custom-registry',
          },
          {
            _hubClusterResource: 'true',
            _ownerUID: 'local-cluster/1234-abcd',
            _uid: 'local-cluster/1234-abcd',
            apiversion: 'v1',
            cluster: 'local-cluster',
            container: 'search-api',
            created: '2024-04-15T14:23:59Z',
            hostIP: '10.0.3.162',
            image: 'quay.io/image',
            kind: 'Pod',
            kind_plural: 'pods',
            label: 'app=search; component=search-v2-operator; name=search-api; pod-template-hash=69775dc595',
            name: 'search-api-69775dc595-lhrk9',
            namespace: 'open-cluster-management',
            podIP: '10.130.0.114',
            restarts: '0',
            startedAt: '2024-04-15T14:23:59Z',
            status: 'Running',
          },
          {
            _hubClusterResource: 'true',
            _ownerUID: 'local-cluster/1234-abcd',
            _uid: 'local-cluster/1234-abcd',
            apiversion: 'v1',
            cluster: 'local-cluster',
            container: 'search-collector',
            created: '2024-04-15T14:23:59Z',
            hostIP: '10.0.68.86',
            image: 'quay.io/image',
            kind: 'Pod',
            kind_plural: 'pods',
            name: 'search-collector-764d748c4f-pbgrb',
            namespace: 'open-cluster-management',
            podIP: '10.129.0.117',
            restarts: '0',
            startedAt: '2024-04-15T14:23:59Z',
            status: 'Running',
          },
        ],
        __typename: 'SearchResult',
      },
    ],
  }

  const searchDefinitions = getSearchDefinitions((key) => key)
  generateSearchResultExport('test-search-export', searchResultDataMock, searchDefinitions, toastContextMock, t)

  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Generating data. Download may take a moment to start.',
    type: 'info',
    autoClose: true,
  })
  expect(window.URL.createObjectURL).toHaveBeenCalledTimes(1)
  expect(toastContextMock.addAlert).toHaveBeenCalledWith({
    title: 'Export successful',
    type: 'success',
    autoClose: true,
  })
})
