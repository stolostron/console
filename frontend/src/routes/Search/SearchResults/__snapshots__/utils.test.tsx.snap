// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly return VirtualMachine with actions disabled 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete VirtualMachine",
  },
]
`;

exports[`Correctly return VirtualMachine with actions enabled 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "startVM",
    "title": "Start VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "stopVM",
    "title": "Stop VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "restartVM",
    "title": "Restart VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "pauseVM",
    "title": "Pause VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "unpauseVM",
    "title": "Unpause VirtualMachine",
  },
  Object {
    "addSeparator": true,
    "click": [Function],
    "id": "edit",
    "title": "Edit VirtualMachine",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete VirtualMachine",
  },
]
`;

exports[`Correctly return empty row Actions for Application 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "view-application",
    "title": "View Application",
  },
  Object {
    "click": [Function],
    "id": "view-application-topology",
    "title": "View Application topology",
  },
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit Application",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete Application",
  },
]
`;

exports[`Correctly return empty row Actions for restricted resource 1`] = `Array []`;

exports[`Correctly return row Actions 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit Pod",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete Pod",
  },
]
`;

exports[`Correctly return row Actions for Application in global search 1`] = `
Array [
  Object {
    "click": [Function],
    "id": "view-application",
    "title": "View Application",
  },
  Object {
    "click": [Function],
    "id": "view-application-topology",
    "title": "View Application topology",
  },
  Object {
    "click": [Function],
    "id": "edit",
    "title": "Edit Application",
  },
  Object {
    "click": [Function],
    "id": "view-related",
    "title": "View related resources",
  },
  Object {
    "click": [Function],
    "id": "delete",
    "title": "Delete Application",
  },
]
`;
